File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
private:
    void bfs(int row,int col,vector<vector<int>> &vis, vector<vector<char>> &grid){
        
        int n=grid.size();
        int m=grid[0].size();
        vis[row][col]=1;
        queue<pair<int,int>> q;
        q.push({row,col});

        while(!q.empty()){
            int row=q.front().first;
            int col=q.front().second;
            q.pop();
            for(int i=-1;i<=1;i++){
                for(int j=-1;j<=1;j++){
                    if (abs(i) + abs(j) != 1) continue;
                    int nrow=row+i;
                    int ncol=col+j;
                    if(nrow>=0 && nrow<n && ncol>=0 && ncol<m 
                    && grid[nrow][ncol]=='1' && !vis[nrow][ncol]){
                            
                            vis[nrow][ncol]=1;
                            q.push({nrow,ncol});
                    }
                }   
            }

        }

         
    }

    void dfs(int row,int col,vector<vector<int>> &vis, vector<vector<char>> &grid){

        int n=grid.size();
        int m=grid[0].size();

        if(row>=0 && row<n && col>=0 && col<m && !vis[row][col] && grid[row][col]=='1'){
            vis[row][col]=1;
            dfs(row+1,col,vis,grid);
            dfs(row-1,col,vis,grid);
            dfs(row,col+1,vis,grid);
            dfs(row,col-1,vis,grid);
        }
        return;

    }
public:
    int numIslands(vector<vector<char>>& grid) {

        int n=grid.size();
        int m=grid[0].size();
        int island=0;
        vector<vector<int>> vis(n,vector<int> (m,0));

        for(int row=0;row<n;row++){
            for(int col=0;col<m;col++){
                if(!vis[row][col] && grid[row][col] == '1'){
                    island++;
                    //bfs(row,col,vis,grid);
                    dfs(row,col,vis,grid);
                }
            }
        }
        return island;
        
    }
};

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """


        n=len(grid)
        m=len(grid[0])

        vis=[[0 for _ in range(m)] for _ in range(n)]

        count = 0    
        for i in range(n):
            for j in range(m):

                if not vis[i][j] and grid[i][j]=='1':

                    count+=1
                    #self.bfs(i,j,vis,grid)
                    self.dfs(i,j,vis,grid)

        return count

    def dfs(self,i,j,vis,grid):

        n=len(grid)
        m=len(grid[0])
        if self.isValid(n,m,i,j) and not vis[i][j] and grid[i][j]=='1':
            vis[i][j]=1
            self.dfs(i+1,j,vis,grid)
            self.dfs(i-1,j,vis,grid)
            self.dfs(i,j+1,vis,grid)
            self.dfs(i,j-1,vis,grid)






    # def bfs(self,i,j,vis,grid):
    #     vis[i][j]=1
    #     q=deque()
    #     q.append((i,j))
    #     n=len(grid)
    #     m=len(grid[0])
    #     while q:
    #         row,col=q.popleft()

    #         for delRow,delCol in ((1,0),(-1,0),(0,1),(0,-1)):

    #             newRow=row+delRow
    #             newCol=col+delCol

    #             if self.isValid(n,m,newRow,newCol) and not vis[newRow][newCol] and grid[newRow][newCol]=='1':
    #                 vis[newRow][newCol]=1
    #                 q.append((newRow,newCol))

    def isValid(self,n,m,row,col):

        if row<0 or row>=n:
            return False

        if col<0 or col>=m:
            return False

        return True